api/controlplane/kubeadm/v1beta2/kubeadm_control_plane_types.go

```diff
@@ -540,31 +524,57 @@ type KubeadmControlPlaneMachineTemplateDeletionSpec struct {
 	NodeDeletionTimeoutSeconds *int32 `json:"nodeDeletionTimeoutSeconds,omitempty"`
 }
 
-// RolloutBefore describes when a rollout should be performed on the KCP machines.
-type RolloutBefore struct {
+// KubeadmControlPlaneRolloutSpec defines the rollout behavior.
+// +kubebuilder:validation:MinProperties=1
+type KubeadmControlPlaneRolloutSpec struct {
+	// before is a field to indicate a rollout should be performed
+	// if the specified criteria is met.
+	// +optional
+	Before KubeadmControlPlaneRolloutBeforeSpec `json:"before,omitempty,omitzero"`
+
+	// after is a field to indicate a rollout should be performed
+	// after the specified time even if no changes have been made to the
+	// KubeadmControlPlane.
+	// Example: In the YAML the time can be specified in the RFC3339 format.
+	// To specify the rolloutAfter target as March 9, 2023, at 9 am UTC
+	// use "2023-03-09T09:00:00Z".
+	// +optional
+	After *metav1.Time `json:"after,omitempty"`
+
+	// strategy specifies how to roll out control plane Machines.
+	// +optional
+	Strategy KubeadmControlPlaneRolloutStrategy `json:"strategy,omitempty,omitzero"`
+}
+
+// KubeadmControlPlaneRolloutBeforeSpec describes when a rollout should be performed on the KCP machines.
+// +kubebuilder:validation:MinProperties=1
+type KubeadmControlPlaneRolloutBeforeSpec struct {
 	// certificatesExpiryDays indicates a rollout needs to be performed if the
 	// certificates of the machine will expire within the specified days.
 	// +optional
-	CertificatesExpiryDays *int32 `json:"certificatesExpiryDays,omitempty"`
+	// +kubebuilder:validation:Minimum=7
```

Comment:
Breaking change? Should also mention this limitation in the godoc

Reply:
No, I moved the validation from the webhook to OpenAPI

Reply:
It was here before: https://github.com/kubernetes-sigs/cluster-api/blob/1cf8729544d0dee3819f399eaf107048651da41c/controlplane/kubeadm/internal/webhooks/kubeadm_control_plane.go#L382